internal/api/cluster.go

```diff
@@ -1,43 +1,86 @@
 package api
 
 import (
-	"encoding/json"
-	"io"
 	"net/http"
 	"strconv"
+	"sync"
 
 	"github.com/gorilla/mux"
 	"github.com/mattermost/mattermost-cloud/internal/model"
 )
 
-// outputJSON is a helper method to write the given data as JSON to the given writer.
-//
-// It only logs an error if one occurs, rather than returning, since there is no point in trying
-// to send a new status code back to the client once the body has started sending.
-func outputJSON(c *Context, w io.Writer, data interface{}) {
-	encoder := json.NewEncoder(w)
-	err := encoder.Encode(data)
+// initCluster registers cluster endpoints on the given router.
+func initCluster(apiRouter *mux.Router, context *Context) {
+	addContext := func(handler contextHandlerFunc) *contextHandler {
+		return newContextHandler(context, handler)
+	}
+
+	clustersRouter := apiRouter.PathPrefix("/clusters").Subrouter()
+	clustersRouter.Handle("", addContext(handleGetClusters)).Methods("GET")
+	clustersRouter.Handle("", addContext(handleCreateCluster)).Methods("POST")
+
+	clusterRouter := apiRouter.PathPrefix("/cluster/{cluster:[A-Za-z0-9]{26}}").Subrouter()
+	clusterRouter.Handle("", addContext(handleGetCluster)).Methods("GET")
+	clusterRouter.Handle("", addContext(handleRetryCreateCluster)).Methods("POST")
+	clusterRouter.Handle("/kubernetes/{version}", addContext(handleUpgradeCluster)).Methods("PUT")
+	clusterRouter.Handle("", addContext(handleDeleteCluster)).Methods("DELETE")
+}
+
+// lockCluster synchronizes access to the given cluster across potentially multiple provisioning
+// servers.
+func lockCluster(c *Context, clusterID string) (*model.Cluster, int, func()) {
+	cluster, err := c.Store.GetCluster(clusterID)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to query cluster")
+		return nil, http.StatusInternalServerError, nil
+	}
+	if cluster == nil {
+		return nil, http.StatusNotFound, nil
+	}
+
+	locked, err := c.Store.LockCluster(clusterID, c.RequestID)
 	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to encode result")
+		c.Logger.WithError(err).Error("failed to lock cluster")
+		return nil, http.StatusInternalServerError, nil
+	} else if !locked {
+		c.Logger.Error("failed to acquire lock for cluster")
+		return nil, http.StatusConflict, nil
+	}
+
+	unlockOnce := sync.Once{}
+
+	return cluster, 0, func() {
+		unlockOnce.Do(func() {
+			unlocked, err := c.Store.UnlockCluster(cluster.ID, c.RequestID, false)
+			if err != nil {
+				c.Logger.WithError(err).Errorf("failed to unlock cluster")
+			} else if unlocked != true {
+				c.Logger.Warn("failed to release lock for cluster")
```

Comment:
This is an error that needs to be dealt with manually (cluster is stuck locked) Error seems appropriate.